### PR TITLE
GH-36399: [Python] Add missing `shape` property to `RecordBatch`

### PR DESCRIPTION
**Pull request by judahrand**
### Rationale for this change

This property is currently missing.

### What changes are included in this PR?

Move the `shape` property to the `_Tabular` base class.

### Are these changes tested?

Yes. I've merged the `test_table_basics` and `test_record_batch` basics tests to make sure that all the same basics are tests on both classes.

### Are there any user-facing changes?

* GitHub Issue: #36399

### CODE REVIEW

File: python/pyarrow/tests/test_table.py [hunk 1]
```diff
@@ -1246,53 +1184,141 @@ def test_table_to_batches():
...
+    table = cls.from_arrays(data, schema=schema)
+    assert table.schema == schema
+
     wr = weakref.ref(table)
     assert wr() is not None
     del table
     assert wr() is None
 
 
+def test_table_str():
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_table.py:1264]: There is already a `test_table_repr_to_string` that should cover this, I think?

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks, just two minor comments

File: python/pyarrow/tests/test_table.py (hunk 1 above)
Comment (judahrand) [reply python/pyarrow/tests/test_table.py]: https://github.com/apache/arrow/pull/40643/commits/1411614f512039d65fef41a835709c977aa2d994

**Review by jorisvandenbossche — APPROVED** [review]
Thanks!

**Pull request merged**